Pushing example to manage downloaded offline regions

ZiZasaurus opened the pull request:
Created an example for parity between iOS and Android regarding managing offline regions

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -0,0 +1,220 @@
+        var packArray = [MGLOfflinePack]()
+        var userInfo = [String]()
+        var pack = MGLOfflineStorage.shared.packs
+        var progressPercentage: Float?
```
Instead of declaring this here, perhaps consider extending `MGLOfflinePackProgress` to have a computed property that returns the progress percentage. 

By doing this, the responsibility of calculating the progress of the pack moves from the `ViewController` class to the `MGLOfflinePackProgress` object itself. 

You don't need to add the extension to another file, just include it at the bottom of the file and make sure its marked as `fileprivate`, so it can't be accidentally used in future examples.

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -0,0 +1,220 @@
+                              zoomLevel: 13, animated: false)
+
+            // Setup offline pack notification handlers.
+            NotificationCenter.default.addObserver(self, selector: #selector(offlinePackProgressDidChange), name: NSNotification.Name.MGLOfflinePackProgressChanged, object: nil)
```
Let's move this out to a function to keep things organized.

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -0,0 +1,220 @@
+
+        func installConstraints() {
+
+            if #available(iOS 11.0, *) {
```
I think these constraints should be relative to the map view, then we could possible get around not having to do this `if #available` check.

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -0,0 +1,220 @@
+
+            view.addSubview(mapView)
+            view.addSubview(tableView)
+            view.addSubview(downloadButton)
```
The download button should be added to the map view, not the root view.

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -0,0 +1,220 @@
+        override func viewDidDisappear(_ animated: Bool) {
+
+            super.viewDidDisappear(animated)
+            // When leaving this view controller, suspend offline downloads.
```
I think we should instead delete all the offline packs, since they will continue to exists in the database forever until the app is deleted/reinstalled.

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -0,0 +1,220 @@
+
+        func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+            let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)
+            if MGLOfflineStorage.shared.packs?.count != nil {
```
I'd `guard` here instead:

```swift
guard let packs = MGLOfflinePacks.shared.packs { else return }

// Access to packs here
```

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -0,0 +1,220 @@
+            let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)
+            if MGLOfflineStorage.shared.packs?.count != nil {
+                let progress = Float(packArray[indexPath.row].progress.countOfResourcesCompleted)/Float(packArray[indexPath.row].progress.countOfResourcesExpected)
+                cell.textLabel!.text = "Region \(indexPath.row + 1): \n  Percent completion: \(String(format: "%.2f", progress * 100)) %"
```
I'd recommend against having a line break here - maybe just a dash instead? Then we don't need to explicitly set the cell height.

captainbarbosa replied:
`cell.textLabel!.text` is force unwrapped here, but not on the other text label properties. I think you're fine just using `cell.textLabel?.text`.

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -0,0 +1,220 @@
+                cell.textLabel!.text = "Region \(indexPath.row + 1): \n  Percent completion: \(String(format: "%.2f", progress * 100)) %"
+                print(Float((packArray[indexPath.row].progress.countOfResourcesCompleted)/(packArray[indexPath.row].progress.countOfResourcesExpected) * 100))
+
+                cell.textLabel?.font = UIFont.systemFont(ofSize: 15)
```
Let's just use the system font size, meaning this line can just be removed.

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -0,0 +1,220 @@
+                cell.textLabel?.numberOfLines = 0
+                cell.textLabel?.lineBreakMode = .byWordWrapping
+                if #available(iOS 13.0, *) {
+                    cell.textLabel?.textColor = .darkGray
```
Let's keep the color the default so we can avoid doing this `if #available` check.

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -0,0 +1,220 @@
+
+            // Create a region that includes the current viewport and any tiles needed to view it when zoomed further in.
+            // Because tile count grows exponentially with the maximum zoom level, you should be conservative with your `toZoomLevel` setting.
+            let region = MGLTilePyramidOfflineRegion(styleURL: mapView.styleURL, bounds: mapView.visibleCoordinateBounds, fromZoomLevel: mapView.zoomLevel, toZoomLevel: 14)
```
What if the user is already past z14, would this break?

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -0,0 +1,220 @@
+
+            // Setup offline pack notification handlers.
+            NotificationCenter.default.addObserver(self, selector: #selector(offlinePackProgressDidChange), name: NSNotification.Name.MGLOfflinePackProgressChanged, object: nil)
+            NotificationCenter.default.addObserver(self, selector: #selector(offlinePackDidReceiveError), name: NSNotification.Name.MGLOfflinePackError, object: nil)
```
I don't think we need to watch these two notification types, since they're covered in the other example. We can just watch `MGLOfflinePackProgressChanged`.

captainbarbosa reviewed (changes requested):
I like how this ended up! Let me know if you have questions 👍

ZiZasaurus commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -0,0 +1,220 @@
+
+@objc(ManageOfflineRegionsExample_Swift)
+
+class ManageOfflineRegionsExample_Swift: UIViewController, MGLMapViewDelegate, UITableViewDelegate, UITableViewDataSource  {
```
added the delete swipe function! :)

ZiZasaurus commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -0,0 +1,220 @@
+
+            if #available(iOS 11.0, *) {
+                NSLayoutConstraint.activate([
+                    downloadButton.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
```
no compass in this example but i'm happy to move over to the left hand side. Whatever you think looks best!

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -1,220 +1,234 @@
+            tableView.trailingAnchor.constraint(equalTo: mapView.trailingAnchor),
+            tableView.heightAnchor.constraint(equalTo: mapView.heightAnchor, multiplier: 0.4),
+            downloadButton.topAnchor.constraint(equalTo: mapView.topAnchor, constant: 10),
+            downloadButton.leadingAnchor.constraint(equalTo: mapView.leadingAnchor),
```
Can we add some extra space here? Looks a little too close to the edge:

![image](https://user-images.githubusercontent.com/10850812/96941790-41624880-1488-11eb-95de-d8b51766fb6b.png)

I'd also consider making the button a contrasting color with some bold text so it stands out from the map view's color scheme. Maybe try one of the [system colors](https://developer.apple.com/documentation/uikit/uicolor/standard_colors)?

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -1,220 +1,234 @@
+            print("so...")
+            cell.textLabel?.text = "No Offline Packs Saved"
+            cell.textLabel?.textColor = .darkGray
+            cell.textLabel?.textAlignment = .center
```
This would look nicer in a header. It would also be more efficient, since `cellForRowAt:indexPath` can potentially be called a lot of you have a lot of cells. If you already have at least one cell, then you know you've got packs, and so the rest of this if/else check is extra work.

`UITableViewDataSource` supports using titles to mark section headers. 

You can try something like:

```swift
func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
   /**
     If there are downloaded packs, return "Offline packs",
     otherwise return "No pack downloaded".
     */
}
```

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -1,220 +1,234 @@
-            return cell
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        var cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)
+        cell = UITableViewCell(style: .subtitle, reuseIdentifier: "cell")
```
`dequeueReusableCell(identifier:)` always returns a cell, so re-assigning it to a cell again is unnecessary.

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -1,220 +1,234 @@
+    }
 
-        @objc func startOfflinePackDownload() {
+    @available(iOS 11.0, *)
```
I didn't realize this was an iOS 11+ things. Let's leave this part out since we don't want to have to deal with many `@available` checks in examples, as its not integral to the example being effective.

captainbarbosa reviewed (commented):
Let's review this in our next meetup

ZiZasaurus commented:
Worked with @captainbarbosa to add extension for progressPercentage calculation -- i added an if statement in the extension because it wouldn't run properly otherwise.

Also addressed the following issues: 
- why cells were duplicating
- cleaned up formatting and removed unnecessary print statements (left the ones from the original example)
- fixed cell subtitle issue

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -111,23 +94,28 @@ class ManageOfflineRegionsExample_Swift: UIViewController, MGLMapViewDelegate, U
     func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
-        var cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)
+
         cell = UITableViewCell(style: .subtitle, reuseIdentifier: "cell")
```
Ohhh see what you mean now with the cell style thing. Here's a simply approach I think could work:

```swift
let cell = tableView.dequeueReusableCell(withIdentifier: "cell") ?? UITableViewCell(style: .subtitle, 
                                                                    reuseIdentifier: "cell")
```

This checks `dequeueReusableCell` to see if there is already a cell prepared with that identifier, and if there isn't, it creates a new cell with the `subtitle` style.

ZiZasaurus replied:
changing the code to this still doesn't allow the subtitle to show :(

ZiZasaurus commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -111,23 +94,28 @@ class ManageOfflineRegionsExample_Swift: UIViewController, MGLMapViewDelegate, U
         }
     }
 
+    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
```
Yea I realized this and forgot to change it *facepalm*

ZiZasaurus commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -221,14 +178,30 @@ class ManageOfflineRegionsExample_Swift: UIViewController, MGLMapViewDelegate, U
+        if countOfResourcesExpected != 0 {
+            return Float((countOfResourcesCompleted / countOfResourcesExpected) * 100)
+        } else {
+            return 1.0
```
changed it to 0

ZiZasaurus commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -221,14 +178,30 @@ class ManageOfflineRegionsExample_Swift: UIViewController, MGLMapViewDelegate, U
                 // Otherwise, print download/verification progress.
 
-                print("Offline pack “\(userInfo["name"] ?? "unknown")” has \(completedResources) of \(expectedResources) resources — \(String(format: "%.2f", progressPercentage! * 100))%.")
+                print("Offline pack “\(userInfo["name"] ?? "unknown")” has \(completedResources) of \(expectedResources) resources — \(String(format: "%.2f", pack.progress.progressPercentage))%.")
```
nope not needed! removed!

ZiZasaurus commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -202,14 +164,9 @@ class ManageOfflineRegionsExample_Swift: UIViewController, MGLMapViewDelegate, U
-
-            progressPercentage = Float(completedResources) / Float(expectedResources)
+            let completedResources = pack.progress.countOfResourcesCompleted
+            let expectedResources = pack.progress.countOfResourcesExpected
```
removed!

ZiZasaurus commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -111,23 +94,28 @@ class ManageOfflineRegionsExample_Swift: UIViewController, MGLMapViewDelegate, U
+
             cell.textLabel?.text = "Region \(indexPath.row + 1): size: \(packs[indexPath.row].progress.countOfBytesCompleted)"
-            cell.detailTextLabel?.text = "Percent completion: \(String(format: "%.2f", progress * 100))%"
+            cell.detailTextLabel?.text = "Percent completion: \(String(format: "%.2f", packs[indexPath.row].progress.progressPercentage))%"
```
fixed!

captainbarbosa commented:
@ZiZasaurus please re-request review when ready (the little 🔁 symbol next to my name in the reviewers list)

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -0,0 +1,190 @@
+
+    @objc func startOfflinePackDownload() {
+
+        // Create a region that includes the current viewport and any tiles needed to view it when zoomed further in.
+        // Because tile count grows exponentially with the maximum zoom level, you should be conservative with your `toZoomLevel` setting.
```
```suggestion
        /**
         Create a region that includes the current map camera, to be captured
         in an offline map. Note: Because tile count grows exponentially as zoom level
         increases, you should be conservative with your `toZoomLevel` setting.
         */
```

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -0,0 +1,190 @@
+        // Because tile count grows exponentially with the maximum zoom level, you should be conservative with your `toZoomLevel` setting.
+        let region = MGLTilePyramidOfflineRegion(styleURL: mapView.styleURL, bounds: mapView.visibleCoordinateBounds, fromZoomLevel: mapView.zoomLevel, toZoomLevel: mapView.zoomLevel + 2)
+
+        // Store some data for identification purposes alongside the downloaded resources.
```
```suggestion
        // Store some data for identification purposes alongside the offline pack.
```

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -0,0 +1,190 @@
+        view.addSubview(mapView)
+        view.addSubview(tableView)
+        mapView.addSubview(downloadButton)
+        mapView.setCenter(CLLocationCoordinate2D(latitude: 22.27933, longitude: 114.16281),
+
+                          zoomLevel: 13, animated: false)
```
```suggestion
        let centerCoordinate = CLLocationCoordinate2D(latitude: 22.27933, longitude: 114.16281)
        mapView.setCenter(centerCoordinate, zoomLevel: 13, animated: false)
```

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -0,0 +1,190 @@
+
+            }
+        }
+        self.tableView.reloadData()
```
```suggestion
        // Reload the table to update the progress percentage for each offline pack.
        self.tableView.reloadData()
```

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -0,0 +1,190 @@
+    }
+
+}
+
```
Let's move all the methods that deal with setting up the `UITableView` here, into an extension that conforms to the protocols it needs:

```swift
extension ManageOfflineRegionsExample_Swift: UITableViewDelegate, UITableViewDataSource {
    // table view methods here
}
```

captainbarbosa reviewed (commented):
I've made some formatting suggestions which you can apply via [GitHub's suggested changes feature](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/incorporating-feedback-in-your-pull-request#applying-a-suggested-change), and have a couple of minor code comments. Please also check line spacing as well - I was seeing a couple of SwiftLint warnings. This is close!

ZiZasaurus commented:
Git made me do a merge so let me know if anything looks off

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -0,0 +1,197 @@
+            mapView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
+            mapView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
+            mapView.heightAnchor.constraint(equalTo: view.heightAnchor, multiplier: 0.75),
+            tableView.topAnchor.constraint(equalTo: mapView.bottomAnchor),
+            tableView.leadingAnchor.constraint(equalTo: mapView.leadingAnchor),
+            tableView.trailingAnchor.constraint(equalTo: mapView.trailingAnchor),
+            tableView.heightAnchor.constraint(equalTo: mapView.heightAnchor, multiplier: 0.4),
+            downloadButton.topAnchor.constraint(equalTo: mapView.topAnchor, constant: 100),
+            downloadButton.leadingAnchor.constraint(equalTo: mapView.leadingAnchor, constant: 5),
+            downloadButton.trailingAnchor.constraint(equalTo: mapView.leadingAnchor, constant: 150)
```
I think there are some improvements that can be made here -

**Map view / table view**:

The only constraints here that should be relative to the `mapView` are the table view's top. The rest can just be relative to the `view` itself.

<details>
<summary>OR, you can take on a ✨ bonus challenge ✨... </summary>

<br>

The map view and table view take up almost 50/50 in height. Instead of manually setting the constrains for these views, I challenge you to experiment with placing both of these in a [`UIStackView`](https://developer.apple.com/documentation/uikit/uistackview). The `UIStackView` will need to have its `distribution` and `axis` properties set to the correct values for this to work.
</details>

**Download button**:

Instead of doing a trailing anchor for the download button, let's set a `width` constraint instead. I'd also add some `height` since the button looks a little squished:

![image](https://user-images.githubusercontent.com/10850812/99131895-3acb7a80-25c9-11eb-8ac4-2a589f125ae8.png)


To fit with the design and to stand out from the map more, can you reverse the button's text color and background color?

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -0,0 +1,197 @@
+    }
+
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        guard let packs = MGLOfflineStorage.shared.packs else { return }
```
Just adding an extra space below this line:

```suggestion
        guard let packs = MGLOfflineStorage.shared.packs else { return }
        
```

captainbarbosa reviewed (changes requested):
Added some formatting suggestions, I think the UI needs a little cleanup and then we should be good to go 🚀 

Also, I'd be fine with the Obj-C version coming as a follow up PR. This PR is already a little big.

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -0,0 +1,220 @@
+
+}
+
+fileprivate extension MGLOfflinePackProgress {
```
Lets move this below the table view methods.

captainbarbosa replied:
(Meaning, move `fileprivate extension MGLOfflinePackProgress {...}` below all the table view methods)

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -0,0 +1,220 @@
+        if let packs = MGLOfflineStorage.shared.packs {
+            let pack = packs[indexPath.row]
+
+            cell.textLabel?.text = "Region \(indexPath.row + 1): size: \(pack.progress.countOfBytesCompleted)"
```
I think we should convert `pack.progress.countOfBytesCompleted` to MB, such that 600000 bytes could be transformed to a string such as `0.6 MB`.

Check out [`ByteCountFormatter`](https://developer.apple.com/documentation/foundation/bytecountformatter) which helps you do this. You'll want to create the `ByteCountFormatter` _outside_ of this method (so in a global constant) so it only gets created once for the entire table instead of for a per-cell basis, and then call the appropriate method to format the value here.

If you don't end up implementing this, I'd at least add the units to the end of this string as the user doesn't know what unit value is being displayed.

captainbarbosa reviewed (approved):
✅ pending some minor comments. Nice work!

captainbarbosa commented on Examples/Swift/ManageOfflineRegionsExample.swift:
```diff
@@ -5,6 +5,8 @@ import Foundation
 
 class ManageOfflineRegionsExample_Swift: UIViewController, MGLMapViewDelegate {
 
+    let inBytes = ByteCountFormatter()
```
I accidentally led you astray with this suggestion, I'm sorry! 

So, it looks like the `ByteCountFormatter.string(fromByteCount:countStyle:)` is a _class_ method. This means we don't need to initialize a `ByteCountFormatter` to call this specific method:

For example,

```swift
ByteCountFormatter().string(fromByteCount:countStyle:) // won't work
ByteCountFormatter.string(fromByteCount:countStyle:) // works
```

Because of this, there actually isn't a need to hold a reference to a `ByteCountFormatter` here, because we don't need an instance of the `ByteCountFormatter` class to call the method we want to use. 

In this file you're using two different methods on `ByteCountFormatter`. One is the class function (or class method), and the other is an instance method:

```swift
// Within offlinePackProgressDidChange
// Class method - called on the type itself
let byteCount = ByteCountFormatter.string(fromByteCount: Int64(pack.progress.countOfBytesCompleted), countStyle: ByteCountFormatter.CountStyle.memory)

// Within cellForRowAt:indexPath
// Instance method - called on an instance of the type
cell.textLabel?.text = "Region \(indexPath.row + 1): size: \(inBytes.string(fromByteCount: Int64(pack.progress.countOfBytesCompleted)))"
```

Here's a suggestion - how about we add a second computed property on `MGLOfflinePackProgress`:

```swift
var formattedCountOfBytesCompleted: String {
    return ByteCountFormatter.string(fromByteCount: Int64(countOfBytesCompleted),
                                     countStyle: .memory)
}
```

And then the updates will look like this:

```swift
// Within offlinePackProgressDidChange
let byteCount = pack.progress.formattedCountOfBytesCompleted

// Within cellForRowAt:indexPath
cell.textLabel?.text = "Region \(indexPath.row + 1): size: \(pack.progress.formattedCountOfBytesCompleted)"
```

captainbarbosa reviewed (approved):
Please merge after reviewing the last suggestion 👍 🏆

captainbarbosa commented:
<img width=260 src="https://user-images.githubusercontent.com/10850812/100126437-cac3bc80-2e32-11eb-83b7-2e229e50d92c.gif">
